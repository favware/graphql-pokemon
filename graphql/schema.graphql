"A single Pokémon ability entry"
type Ability {
  "The key of the ability as stored in the API"
  key: AbilitiesEnum!

  "Bulbapedia page for an ability"
  bulbapediaPage: String!

  "The long description for an ability"
  desc: String

  "Whether this ability has effects outside of battle, and if so what the effect is"
  isFieldAbility: String

  "The name for an ability"
  name: String!

  """
  The Pokémon that have this ability.
  Note that when querying this field and nesting deep into Pokemon.abilities, that the nested list
  will not have any values to prevent infinite looping data.
  """
  pokemonThatHaveThisAbility: [Pokemon]!

  "Serebii page for an ability"
  serebiiPage: String!

  "The short description for an ability"
  shortDesc: String!

  "Smogon page for an ability"
  smogonPage: String!
}

"A Pokémon's entry"
type Pokemon {
  "The key of the Pokémon as stored in the API"
  key: PokemonEnum!

  "The abilities for a Pokémon"
  abilities: Abilities!

  "The back sprite for a Pokémon. For most Pokémon this will be the animated gif, with some exceptions that were older-gen exclusive"
  backSprite: String!

  "Base form if this entry describes an alternate form"
  baseForme: String

  "Base species if this entry describes a special form"
  baseSpecies: String

  "Base stats for a Pokémon"
  baseStats: Stats!

  "The total of all base stats for a Pokémon"
  baseStatsTotal: Int!

  "Bulbapedia page for a Pokémon"
  bulbapediaPage: String!

  "The catch rate data for a Pokémon"
  catchRate: CatchRate

  "The colour of a Pokémon as listed in the Pokedex"
  color: String!

  "Any other *cosmetic* forms for a Pokémon, distinguished from other formes as cosmetic formes only change the look of the Pokémon, while other formes might also change an ability, move set or other data."
  cosmeticFormes: [String!]

  "The egg groups a Pokémon is in"
  eggGroups: [String!]

  "The evolution level, or special method, for a Pokémon"
  evolutionLevel: String

  "The evolutions for a Pokémon, if any"
  evolutions: [Pokemon!]

  "EV yields for a Pokémon"
  evYields: EvYields!

  "The flavor texts for a Pokémon"
  flavorTexts: [Flavor!]!

  "The form identifier of a Pokémon"
  forme: String

  "The single letter identifier of the form"
  formeLetter: String

  "The gender data for a Pokémon"
  gender: Gender!

  "The height of a Pokémon in meters"
  height: Float!

  "Whether the egg of a Pokémon is obtainable"
  isEggObtainable: Boolean!

  "The levelling rate of a Pokémon"
  levellingRate: String

  "The maximum number of steps required for the egg of a Pokémon to hatch"
  maximumHatchTime: Int

  "The minimum number of steps required for the egg of a Pokémon to hatch"
  minimumHatchTime: Int

  "The dex number for a Pokémon"
  num: Int!

  "The learnset for this pokemon"
  learnsets: GenerationalPokemonLearnset

  "Any other forms for a Pokémon"
  otherFormes: [String!]

  "The preevolutions for a Pokémon, if any"
  preevolutions: [Pokemon!]

  "Serebii page for a Pokémon"
  serebiiPage: String!

  "The shiny back sprite for a Pokémon. For most Pokémon this will be the animated gif, with some exceptions that were older-gen exclusive"
  shinyBackSprite: String!

  "The shiny sprite for a Pokémon. For most Pokémon this will be the animated gif, with some exceptions that were older-gen exclusive"
  shinySprite: String!

  "Smogon page for a Pokémon"
  smogonPage: String!

  "The smogon tier a Pokémon falls under"
  smogonTier: String!

  "The species name for a Pokémon"
  species: String!

  "The sprite for a Pokémon. For most Pokémon this will be the animated gif, with some exceptions that were older-gen exclusive"
  sprite: String!

  "The types for a Pokémon"
  types: [PokemonType!]!

  "The weight of a Pokémon in kilograms"
  weight: Float!
}

"The type of a Pokémon and that types matchup"
type PokemonType {
  "The name of the typ"
  name: String!

  "The type matchup for this type"
  matchup: TypeMatchup!
}

"A Pokémon's abilities entry"
type Abilities {
  "The first ability of a Pokémon"
  first: Ability!

  "The hidden ability of a Pokémon"
  hidden: Ability

  "The second ability of a Pokémon"
  second: Ability

  "The special ability of a Pokémon"
  special: Ability
}

"A Pokémon's stats"
type Stats {
  "The base attack stat of a Pokémon"
  attack: Int!

  "The base defense stat of a Pokémon"
  defense: Int!

  "The base HP stat of a pokémon"
  hp: Int!

  "The base special attack stat of a Pokémon"
  specialattack: Int!

  "The base special defense stat of a Pokémon"
  specialdefense: Int!

  "The base speed stat of a Pokémon"
  speed: Int!
}

"A Pokémon catch rate entry"
type CatchRate {
  "The base catch rate for a Pokémon that will be used to calculate the final catch rate"
  base: Int!

  "The chance to capture a Pokémon when an ordinary Poké Ball is thrown at full health without any status condition"
  percentageWithOrdinaryPokeballAtFullHealth: String!
}

"A Pokémon's EV yields"
type EvYields {
  "The attack EV yield of a Pokémon"
  attack: Int!

  "The defense EV yield of a Pokémon"
  defense: Int!

  "The HP EV yield of a pokémon"
  hp: Int!

  "The special attack EV yield of a Pokémon"
  specialattack: Int!

  "The special defense EV yield of a Pokémon"
  specialdefense: Int!

  "The speed EV yield of a Pokémon"
  speed: Int!
}

"A flavor text entry for a Pokémon"
type Flavor {
  "The flavor text for this entry"
  flavor: String!

  "The name of the game this flavor text is from"
  game: String!
}

"A Pokémon gender ratio entry"
type Gender {
  "The percentage for female occurrences"
  female: String!

  "The percentage of male occurrences"
  male: String!
}

"A single item entry"
type Item {
  "The key of the item as stored in the API"
  key: ItemsEnum!

  "Bulbapedia page for an item"
  bulbapediaPage: String!

  "The long description for an item"
  desc: String!

  "The generation in which this item was introduced"
  generationIntroduced: Int!

  "Whether an item is non-standard, and if it is why"
  isNonstandard: String

  "The name for an item"
  name: String!

  "Serebii page for an item"
  serebiiPage: String!

  "The long description for an item"
  shortDesc: String

  "Smogon page for an item"
  smogonPage: String

  "The sprite for an item"
  sprite: String!
}

"The learnset for each Pokémon split by generation"
type GenerationalPokemonLearnset {
  "The learnset of this Pokémon in Generation 8"
  generation8: PokemonLearnset!
  "The learnset of this Pokémon in Generation 7"
  generation7: PokemonLearnset!
  "The learnset of this Pokémon in Generation 6"
  generation6: PokemonLearnset!
  "The learnset of this Pokémon in Generation 5"
  generation5: PokemonLearnset!
  "The learnset of this Pokémon in Generation 4"
  generation4: PokemonLearnset!
  "The learnset of this Pokémon in Generation 3"
  generation3: PokemonLearnset!
}

type PokemonLearnset {
  "The moves that are exclusively learned in the Unova Dream World"
  dreamworldMoves: [LearnsetMove!]

  "The moves that can be passed as egg moves"
  eggMoves: [LearnsetMove!]

  "The moves that are exclusive to event variants of the Pokémon"
  eventMoves: [LearnsetMove!]

  "The moves that can be learned through levelling up"
  levelUpMoves: [LearnsetLevelUpMove!]

  "The moves that can be learned from a Technical Machine or Technical Record"
  tmMoves: [LearnsetMove!]

  "The moves that can be learned from a move tutor"
  tutorMoves: [LearnsetMove!]

  "The moves that can be learned through virtual console transfer"
  virtualTransferMoves: [LearnsetMove!]
}

"A learnset entry"
type Learnset {
  "The back sprite for a Pokémon. For most Pokémon this will be the animated gif, with some exceptions that were older-gen exclusive"
  backSprite: String!

  "The PokéDex colour for the Pokémon"
  color: String!

  "The key of the Pokémon as stored in the API"
  pokemonKey: PokemonEnum!

  "The dex number for a Pokémon"
  num: Int!

  "The shiny back sprite for a Pokémon. For most Pokémon this will be the animated gif, with some exceptions that were older-gen exclusive"
  shinyBackSprite: String!

  "The shiny sprite for a Pokémon. For most Pokémon this will be the animated gif, with some exceptions that were older-gen exclusive"
  shinySprite: String!

  "The species name for a Pokémon"
  species: String!

  "The sprite for a Pokémon. For most Pokémon this will be the animated gif, with some exceptions that were older-gen exclusive"
  sprite: String!

  "The moves that are exclusively learned in the Unova Dream World"
  dreamworldMoves: [LearnsetMove!]

  "The moves that can be passed as egg moves"
  eggMoves: [LearnsetMove!]

  "The moves that are exclusive to event variants of the Pokémon"
  eventMoves: [LearnsetMove!]

  "The moves that can be learned through levelling up"
  levelUpMoves: [LearnsetLevelUpMove!]

  "The moves that can be learned from a Technical Machine or Technical Record"
  tmMoves: [LearnsetMove!]

  "The moves that can be learned from a move tutor"
  tutorMoves: [LearnsetMove!]

  "The moves that can be learned through virtual console transfer"
  virtualTransferMoves: [LearnsetMove!]
}

"A learnset move entry"
type LearnsetMove {
  "The generation in which this pokémon learned the move this way"
  generation: Int!

  "The move"
  move: Move!
}

"A learnset level up move entry"
type LearnsetLevelUpMove {
  "The generation in which this pokémon learned the move this way"
  generation: Int!

  "The move"
  move: Move!

  "The level at which the move is learned"
  level: Int!
}

"A single Pokémon move entry"
type Move {
  "The key of the move as stored in the API"
  key: MovesEnum!

  "The accuracy for a move"
  accuracy: Int!

  "The base power for a move"
  basePower: String!

  "Bulbapedia page for a move"
  bulbapediaPage: String!

  "The category for a move"
  category: String!

  "The contest type for a move"
  contestType: String

  "The long description for a move"
  desc: String

  "Whether this move can be used outside of battle, and if it can what the effect of the field move is"
  isFieldMove: String

  "Whether this move is a G-MAX move, and if it is which Gigantamaxed Pokémon can use it"
  isGMax: String

  "Whether a move is non-standard, and if it is why"
  isNonstandard: String

  "Whether this move is a Z-Move, and if it is the Z-Crystal required to trigger it"
  isZ: String

  "The power this move will have when used with its Max Move equivalent"
  maxMovePower: Int

  "The name for a move"
  name: String!

  "The power points for a move"
  pp: Int!

  "The priority for a move"
  priority: Int!

  "Serebii page for a move"
  serebiiPage: String!

  "The short description for a move"
  shortDesc: String!

  "Smogon page for a move"
  smogonPage: String!

  "The target for a move"
  target: String!

  "The type for a move"
  type: String!

  "The power this move will have when used with its Z-move equivalent"
  zMovePower: Int!
}

"The type matchups for any one or two given types"
type TypeMatchup {
  "The type matchups when attacking"
  attacking: TypeEffectiveness!

  "The type matchups when defending"
  defending: TypeEffectiveness!
}

"A type matchup entry"
type TypeEffectiveness {
  "The types with 4x effectiveness"
  doubleEffectiveTypes: [String!]!

  "The types with 0.25x effectiveness"
  doubleResistedTypes: [String!]!

  "The types with 2x effectiveness"
  effectiveTypes: [String!]!

  "The types with 0x effectiveness"
  effectlessTypes: [String!]!

  "The types with 1x effectiveness"
  normalTypes: [String!]!

  "The types with 0.5x effectiveness"
  resistedTypes: [String!]!
}
