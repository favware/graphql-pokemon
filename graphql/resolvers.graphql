type Query {
  "Gets the details on a Pokémon ability, using the ability name"
  getAbility("The ability to look up" ability: AbilitiesEnum!): Ability!

  """
  Gets details on a Pokémon ability, using a fuzzy search on name

  This can be used to find multiple results based on the query

  You can provide `take` to limit the amount of abilities to return (default: 1), set the offset of where to start with `offset`, and reverse the entire array with `reverse`.
  """
  getFuzzyAbility(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The ability to fuzzily search"
    ability: String!
  ): [Ability!]!

  """
  Gets details on a single Pokémon based on National Pokédex number

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  **Reversal is applied before pagination!**
  """
  getPokemonByDexNumber(
    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true

    "The International PokéDex number of the Pokémon to look up"
    number: Int!
  ): Pokemon!

  """
  Gets details on a single Pokémon based on species name

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  **Reversal is applied before pagination!**
  """
  getPokemon(
    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true

    "The Pokémon to look up"
    pokemon: PokemonEnum!
  ): Pokemon!

  """
  Gets details on one or more Pokémon based on species name

  You can provide `take` to limit the amount of Pokémon to return (default: 1), set the offset of where to start with `offset`, and reverse the entire array with `reverse`.

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  **Reversal is applied before pagination!**
  """
  getFuzzyPokemon(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The Pokémon to fuzzily search"
    pokemon: String!

    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true
  ): [Pokemon!]!

  """
  Returns a list of all the known Pokémon.

  For every Pokémon all the data on each requested field is returned.

  **_NOTE:_ To skip all CAP Pokémon, PokéStar Pokémon, and Missingno provide an `offset` of 87**

  You can provide `take` to limit the amount of Pokémon to return (default: 1), set the offset of where to start with `offset`, and reverse the entire array with `reverse`.

  You can provide `takeFlavorTexts` to limit the amount of flavour texts to return, set the offset of where to start with `offsetFlavorTexts`, and reverse the entire array with `reverseFlavorTexts`.

  While the API will currently not rate limit the usage of this query, it may do so in the future.

  It is advisable to cache responses of this query.
  """
  getAllPokemon(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1273

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "Sets the offset for flavor texts from where to start."
    offsetFlavorTexts: Int = 0

    "Return this many flavour texts, up to the maximum of entries that the requested Pokémon has."
    takeFlavorTexts: Int = 1

    "Whether to reverse the list of games from which to get the data. By default Generation 1 is considered for `take` first, when setting this to true that is instead Generation 8."
    reverseFlavorTexts: Boolean = true
  ): [Pokemon!]!

  "Gets the details on a Pokémon item, using the item name"
  getItem("The item to look up" item: ItemsEnum!): Item!

  """
  Gets details on a Pokémon item, using a fuzzy search on name

  This can be used to find multiple results based on the query

  By default only 1 result is returned. You can provide the arguments `take`, `offset`, and `reverse` to modify this behaviour.
  """
  getFuzzyItem(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The item to fuzzily search"
    item: String!
  ): [Item!]!

  """
  Gets the learnsets for a given Pokémon and move.

  Multiple moves are possible by putting them in an array: `[move1, move2]`.

  You can also apply a generation filter (only results for the given generation will be returned) with the generation argument
  """
  getLearnset(
    "The generation filter to apply"
    generation: Int

    "The moves to match against the Pokémon"
    moves: [MovesEnum!]!

    "The Pokémon for which to get the learnset"
    pokemon: PokemonEnum!
  ): Learnset!

  """
  Gets the learnset for a given Pokémon and move.

  A fuzzy search is performed to find a matching Pokémon and move

  Multiple moves are possible by putting them in an array: `[move1, move2]`.

  You can also apply a generation filter (only results for the given generation will be returned) with the generation argument
  """
  getFuzzyLearnset(
    "The generation filter to apply"
    generation: Int

    "The moves to match against the Pokémon"
    moves: [String!]!

    "The Pokémon for which to get the learnset"
    pokemon: String!
  ): Learnset!

  "Gets the details on a Pokémon move, using the move name"
  getMove("The move to look up" move: MovesEnum!): Move!

  """
  Gets details on a Pokémon move, using a fuzzy search on name

  This can be used to find multiple results based on the query

  By default only 1 result is returned. You can provide the arguments `take`, "offset", and "reverse" to modify this behaviour.
  """
  getFuzzyMove(
    "Sets the offset where to start"
    offset: Int = 0

    "Return only this many results, starting from the offset"
    take: Int = 1

    "Reverses the dataset before paginating"
    reverse: Boolean = false

    "The move to fuzzily search"
    move: String!
  ): [Move!]!

  "Gets the type matchup data for the given type or types"
  getTypeMatchup("The primary type to check" primaryType: TypesEnum!, "The secondary type to check" secondaryType: TypesEnum): TypeMatchup!
}
