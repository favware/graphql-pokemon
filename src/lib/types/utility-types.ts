/**
 * Removes `null`, `undefined` and `?` from an object
 * This essentially combines `Required<T>`,
 * `NonNullable<T>` and `NonUndefined<T>`
 */
export type NonNullish<T> = {
  [P in keyof T]-?: Exclude<T[P], undefined | null>;
};

export type UnwrapArray<T extends unknown[]> = T extends (infer U)[] ? U : never;

export type Generation = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;
