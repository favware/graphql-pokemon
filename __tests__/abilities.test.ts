import { formatResponse, gCall } from './testUtils/testUtils';
import { DataResponse } from './testUtils/types';
import { GraphQLError } from 'graphql';

describe('getAbilityDetailsByName', () => {
  const getAbilityDetailsByName = `
  query ($ability: Abilities!) {
    getAbilityDetailsByName(ability: $ability) {
      name
    }
  }`;

  test('GIVEN a valid ability THEN returns AbilityEntry', async () => {
    const { data } = await gCall({
      source: getAbilityDetailsByName,
      variableValues: { ability: 'multiscale' },
    }).then(formatResponse) as DataResponse<'getAbilityDetailsByName'>;

    expect(data.getAbilityDetailsByName.name).toBe('Multiscale');
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid ability THEN returns error', async () => {
    const data = await gCall({
      source: getAbilityDetailsByName,
      variableValues: { ability: 'totally_invalid_ability' },
    });

    data.errors!.forEach(error => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });
});

describe('getAbilityDetailsByFuzzy', () => {
  const getAbilityDetailsByFuzzy = `
  query ($ability: String! $skip: Int $take: Int $reverse: Boolean) {
    getAbilityDetailsByFuzzy(ability: $ability skip: $skip take: $take reverse: $reverse) {
      name
    }
  }`;

  test('GIVEN a valid ability THEN returns AbilityEntry', async () => {
    const { data } = await gCall({
      source: getAbilityDetailsByFuzzy,
      variableValues: { ability: 'multiscaale' },
    }).then(formatResponse) as DataResponse<'getAbilityDetailsByFuzzy'>;

    expect(data.getAbilityDetailsByFuzzy.name).toBe('Multiscale');
    expect(data).toMatchSnapshot();
  });

  test('GIVEN a fuzzy ability THEN returns AbilityEntry', async () => {
    const { data } = await gCall({
      source: getAbilityDetailsByFuzzy,
      variableValues: { ability: 'multi' },
    }).then(formatResponse) as DataResponse<'getAbilityDetailsByFuzzy'>;

    expect(data.getAbilityDetailsByFuzzy.name).toBe('Multiscale');
    expect(data).toMatchSnapshot();
  });

  test('GIVEN a fuzzy ability and pagination THEN returns AbilityEntry', async () => {
    const { data } = await gCall({
      source: getAbilityDetailsByFuzzy,
      variableValues: {
        ability: 'multi',
        skip: 1,
        take: 1,
        reverse: false,
      },
    }).then(formatResponse) as DataResponse<'getAbilityDetailsByFuzzy'>;

    expect(data.getAbilityDetailsByFuzzy.name).toBe('Multitype');
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid ability THEN returns error', async () => {
    const data = await gCall({
      source: getAbilityDetailsByFuzzy,
      variableValues: { ability: 'totally_invalid_ability' },
    });

    data.errors!.forEach(error => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid ability and pagination THEN returns error', async () => {
    const data = await gCall({
      source: getAbilityDetailsByFuzzy,
      variableValues: {
        ability: 'totally_invalid_ability',
        skip: 0,
        take: 1,
        reverse: false,
      },
    });

    data.errors!.forEach(error => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });
});

describe('getAbilityByFuzzy', () => {
  const getAbilityByFuzzy = `
  query ($ability: String! $skip: Int $take: Int $reverse: Boolean) {
    getAbilityByFuzzy(ability: $ability skip: $skip take: $take reverse: $reverse)
  }`;

  test('GIVEN a valid ability THEN returns JSONObject', async () => {
    const { data } = await gCall({
      source: getAbilityByFuzzy,
      variableValues: { ability: 'multiscale' },
    }).then(formatResponse) as DataResponse<'getAbilityByFuzzy'>;

    expect(data).toMatchSnapshot();
  });

  test('GIVEN a fuzzy ability THEN returns JSONObject', async () => {
    const { data } = await gCall({
      source: getAbilityByFuzzy,
      variableValues: { ability: 'multi' },
    }).then(formatResponse) as DataResponse<'getAbilityByFuzzy'>;

    expect(data).toMatchSnapshot();
  });

  test('GIVEN a fuzzy ability and pagination THEN returns JSONObject', async () => {
    const { data } = await gCall({
      source: getAbilityByFuzzy,
      variableValues: {
        ability: 'multi',
        skip: 1,
        take: 1,
        reverse: true,
      },
    }).then(formatResponse) as DataResponse<'getAbilityByFuzzy'>;

    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid ability THEN returns empty array', async () => {
    const { data } = await gCall({
      source: getAbilityByFuzzy,
      variableValues: { ability: 'totally_invalid_ability' },
    }).then(formatResponse) as DataResponse<'getAbilityByFuzzy'>;

    expect(data.getAbilityByFuzzy).toStrictEqual([]);
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid ability and pagination THEN returns empty array', async () => {
    const { data } = await gCall({
      source: getAbilityByFuzzy,
      variableValues: {
        ability: 'totally_invalid_ability',
        skip: 1,
        take: 2,
        reverse: false,
      },
    }).then(formatResponse) as DataResponse<'getAbilityByFuzzy'>;

    expect(data.getAbilityByFuzzy).toStrictEqual([]);
    expect(data).toMatchSnapshot();
  });
});

describe('getAbilityByName', () => {
  const getAbilityByName = `
  query ($ability: Abilities!) {
    getAbilityByName(ability: $ability)
  }`;

  test('GIVEN a valid ability THEN returns JSONObject', async () => {
    const { data } = await gCall({
      source: getAbilityByName,
      variableValues: { ability: 'multiscale' },
    }).then(formatResponse) as DataResponse<'getAbilityByName'>;

    expect(data.getAbilityByName.name).toBe('Multiscale');
    expect(data).toMatchSnapshot();
  });

  test('GIVEN an invalid ability THEN returns error', async () => {
    const data = await gCall({
      source: getAbilityByName,
      variableValues: { ability: 'totally_invalid_ability' },
    });

    data.errors!.forEach(error => expect(error).toBeInstanceOf(GraphQLError));
    expect(data).toMatchSnapshot();
  });
});