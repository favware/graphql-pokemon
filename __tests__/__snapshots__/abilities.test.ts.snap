// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAbilityByFuzzy GIVEN a fuzzy ability THEN returns JSONObject 1`] = `
Object {
  "getAbilityByFuzzy": Array [
    Object {
      "item": Object {
        "name": "Multiscale",
        "shortDesc": "If this Pokémon is at full HP, damage taken from attacks is halved.",
      },
      "refIndex": 134,
    },
    Object {
      "item": Object {
        "name": "Multitype",
        "shortDesc": "If this Pokémon is an Arceus, its type changes to match its held Plate or Z-Crystal.",
      },
      "refIndex": 135,
    },
  ],
}
`;

exports[`getAbilityByFuzzy GIVEN a fuzzy ability and pagination THEN returns JSONObject 1`] = `
Object {
  "getAbilityByFuzzy": Array [
    Object {
      "item": Object {
        "name": "Multiscale",
        "shortDesc": "If this Pokémon is at full HP, damage taken from attacks is halved.",
      },
      "refIndex": 134,
    },
  ],
}
`;

exports[`getAbilityByFuzzy GIVEN a valid ability THEN returns JSONObject 1`] = `
Object {
  "getAbilityByFuzzy": Array [
    Object {
      "item": Object {
        "name": "Multiscale",
        "shortDesc": "If this Pokémon is at full HP, damage taken from attacks is halved.",
      },
      "refIndex": 134,
    },
  ],
}
`;

exports[`getAbilityByFuzzy GIVEN an invalid ability THEN returns empty array 1`] = `
Object {
  "getAbilityByFuzzy": Array [],
}
`;

exports[`getAbilityByFuzzy GIVEN an invalid ability and pagination THEN returns empty array 1`] = `
Object {
  "getAbilityByFuzzy": Array [],
}
`;

exports[`getAbilityByName GIVEN a valid ability THEN returns JSONObject 1`] = `
Object {
  "getAbilityByName": Object {
    "name": "Multiscale",
    "shortDesc": "If this Pokémon is at full HP, damage taken from attacks is halved.",
  },
}
`;

exports[`getAbilityByName GIVEN an invalid ability THEN returns error 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$ability" got invalid value "totally_invalid_ability"; Value "totally_invalid_ability" does not exist in "Abilities" enum.],
  ],
}
`;

exports[`getAbilityDetailsByFuzzy GIVEN a fuzzy ability THEN returns AbilityEntry 1`] = `
Object {
  "getAbilityDetailsByFuzzy": Object {
    "name": "Multiscale",
  },
}
`;

exports[`getAbilityDetailsByFuzzy GIVEN a fuzzy ability and pagination THEN returns AbilityEntry 1`] = `
Object {
  "getAbilityDetailsByFuzzy": Object {
    "name": "Multitype",
  },
}
`;

exports[`getAbilityDetailsByFuzzy GIVEN a valid ability THEN returns AbilityEntry 1`] = `
Object {
  "getAbilityDetailsByFuzzy": Object {
    "name": "Multiscale",
  },
}
`;

exports[`getAbilityDetailsByFuzzy GIVEN an invalid ability THEN returns error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Failed to get data for ability: totally_invalid_ability],
  ],
}
`;

exports[`getAbilityDetailsByFuzzy GIVEN an invalid ability and pagination THEN returns error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Failed to get data for ability: totally_invalid_ability],
  ],
}
`;

exports[`getAbilityDetailsByName GIVEN a valid ability THEN returns AbilityEntry 1`] = `
Object {
  "getAbilityDetailsByName": Object {
    "name": "Multiscale",
  },
}
`;

exports[`getAbilityDetailsByName GIVEN an invalid ability THEN returns error 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$ability" got invalid value "totally_invalid_ability"; Value "totally_invalid_ability" does not exist in "Abilities" enum.],
  ],
}
`;
